scalyr_agent/tests/monitors_manager_test.py

```diff
@@ -74,7 +83,146 @@ def test_module_with_id(self):
+        1 monitor returns non-empty frags. 1 returns None.
+        We then verify that the callback is invoked correctly, only if the fragments change.
+
+        Notes:
+        - FakeClocks are used to avoid having to wait for thread loops.
+        - A FakeAgentLogger is defined so we can modify local test state
+        - Even though this test could have been broken up into 2 smaller ones, it is kept as one to minimize overhead
+            time taken by stop_manager()
+        """
+        counter = Counter()
+        test_frag = 'some_frag'
+        poll_interval = 30
+
+        # Create MonitorsManager + 2 monitors. Set each to daemon otherwise unit test doesn't terminate
+        test_manager = self.__create_test_instance([
+            {
+                'module': 'scalyr_agent.builtin_monitors.test_monitor',
+                'gauss_mean': 0,
+            },
+            {
+                'module': 'scalyr_agent.builtin_monitors.test_monitor',
+                'gauss_mean': 0,
+            },
+            {
+                'module': 'scalyr_agent.builtin_monitors.test_monitor',
+                'gauss_mean': 0,
+            },
+        ], [], extra_toplevel_config={'user_agent_refresh_interval': poll_interval})
+        self.assertEquals(test_manager._user_agent_refresh_interval, 30)  # ensure config setting works
+        self.assertEquals(len(test_manager.monitors), 3)
+        patched_monitor_0 = test_manager.monitors[0]
+        patched_monitor_1 = test_manager.monitors[1]
+        unpatched_monitor = test_manager.monitors[2]
+
+        # Fake the clock to fast-forward MonitorsManager sleep loop
+        fake_clock = scalyr_util.FakeClock()
+        for obj in [test_manager, patched_monitor_0, unpatched_monitor]:
+            obj._run_state = scalyr_util.RunState(fake_clock=fake_clock)
+
+        class FragmentPolls(object):
+            def __init__(self):
+                self.__polls = 0
+                self.__lock = threading.Lock()
+
+            def count(self):
+                self.__lock.acquire()
+                try:
+                    return self.__polls
+                finally:
+                    self.__lock.release();
+
+            def increment(self):
+                self.__lock.acquire()
+                try:
+                    self.__polls += 1
+                finally:
+                    self.__lock.release();
+
+            def poll_until_target(self, target_polls, maxwait):
+                """Blocks until n fragment polls are simulated by the MonitorsManager, or an absolute cutoff time.
+                whichever comes first.
+
+                @param target_polls: Number of polls to reach
+                @param maxwait: Time (seconds) to wait for target to be reached
+                @return: True if number of polls reaches target_polls, else False
+                """
+                deadline = time.time() + maxwait
+                while self.count() < target_polls and time.time() < deadline:
```

Comment:
Sorry, I should have been a little more clear in my last review.  I think you should be using a `Condition` here to wait until the count has reached `target_polls`.  Take a look at the documentation: 
 https://docs.python.org/2/library/threading.html#condition-objects

Essentially, you can use the `wait` call on the condition to wait until the count has reached the right level.  Essentially, every call that increments the count should notify on the on the condition object.  Then, here, first see if the level has been reached, and if it has not, then invoke `wait` with the `maxwait` time, looping until `target_polls` has been reached or the timeout has occurred.

IMO, you should not need to invoke `advance_time` here.  You really should just have to advance the time to poll interval outside of the loop.. and then wait for the count to catch up.  If you find that doesn't work, let's discuss.

Reply:
Hmm, I believe the approach you're mentioning here is exactly what  I tried to implement (after the initial review). I used a condition (instead of lock), then in this loop I wait for the count, up until a maximum time.

The problem I encountered is this: Which thread calls `fake_clock.advance_time`?

Here are the threads I'm aware of:
1. MonitorsManager thread
2. Individual Monitor threads
3. Main (test) thread

The MonitorsManager and individual monitor threads are configured to use the fake clock. They each go to sleep at the end of their main loops and rely on an external thread to advance the fake clock (in order to be awakened).

That leaves the main thread to advance fake clock. So if we use a condition in this here and go to sleep (by calling wait()), nobody is advancing the fake clock, hence MonitorsManager never wakes up to poll and the test always fails.

I then created yet another thread to drive the fake clock, but then realized this loop itself could drive the fake clock. I then realized the condition was overkill and so replaced it with a simple lock. 

Finally, the time.sleep(0.01) isn't absolutely necessary but I put it there just so the testrunner doesn't completely hog 100% CPU in the event of some weird situation where it fails but the maxwait was accidentally set too far out.

I hope this all makes sense. I'm having trouble seeing how I can switch this back to a condition without creating yet new thread which I feel makes this more complex.

Reply:
Sorry it took a bit for me to get back to this.  Yup, I understand what you are talking about in terms of the difficulty of advancing the clock just once.  Yes, to fix this the "right" way would involve too much effort.

What's making me cringe a bit about the current approach is that you are doing a busy wait loop.. and just relying on that sleep 0.01 call giving the monitor thread time to wake up and do the poll.

You could change this loop to rely on the condition variable and get rid of the busy wait loop.  The trick is that you have to make sure the monitor thread is asleep before you advance the clock.  It would look something like this:

```
   while self.count() < target_polls and time.time() < deadline
       fake_clock.block_until_n_waiting_threads(1)    # I think you just need the 1 for the monitor thread
       fake_clock.advance_time(increment_by=poll_interval)
       wait_for_poll_increment(timeout=deadline - time.time()
```

The `wait_for_poll_increment` would be implemented using the condition variable approach with the cv on the poll count.

Does that make sense?  I think that would work.  I would advocate taking the time to implement it that way because it will help set precedents in the code base to use things like `block_until_n_waiting_threads` instead of relying on busy wait loops.

Reply:
Makes sense. Agree with setting precedents. I'll give it a shot!

Reply:
@czerwingithub : I've incorporated your suggestions. Hopefully this LGTY.